rust/src/smtp/smtp.rs

```diff
@@ -0,0 +1,165 @@
+                let cr_idx = self.ts_buf.len() - 2;
+                if !self.ts_buf.is_empty() && self.ts_buf[cr_idx] == 0x0d {
+                    let _ = self.ts_buf.split_off(cr_idx);
+                    // TODO set delimeter len to be used later
+                } else {
+                    self.ts_buf.pop();
+                    // TODO set delimeter len to be used later
+                }
+                self.cur_line = self.ts_buf.clone();
+            } else { // No fragments
+                self.cur_line = input.to_vec();
+                if idx != 0 && input[(idx - 1) as usize] == 0x0d {
+                    self.cur_line.pop();
+                    // TODO set delim for later
+                } else {
+                    // TODO set delim for later
+                }
+            }
+        } else {
+            // Possibly fragmented data, add it to buffer
+            self.ts_buf.extend_from_slice(input);
+            return -1
+        }
+        0
+    }
+
+    pub fn get_line_tc(&mut self, input: &[u8]) -> i8
+    {
+        // TODO Clear parser after integration
+        let lf_idx = input.iter().position(|&r| r == 0x0a);
+        if let Some(idx) = lf_idx {
+            let idx = idx as u8;
+            // TODO Set var here to indicate buffer needs clearning on next iteration
+            if !self.tc_buf.is_empty() { // Last of the fragmented data
+                self.tc_buf.extend_from_slice(input);
+                let cr_idx = self.tc_buf.len() - 2;
+                if !self.tc_buf.is_empty() && self.tc_buf[cr_idx] == 0x0d {
+                    let _ = self.tc_buf.split_off(cr_idx);
+                    // TODO set delimeter len to be used later
+                } else {
+                    self.tc_buf.pop();
+                    // TODO set delimeter len to be used later
+                }
+                self.cur_line = self.tc_buf.clone();
+            } else { // No fragmentc
+                self.cur_line = input.to_vec();
+                if idx != 0 && input[(idx - 1) as usize] == 0x0d {
+                    self.cur_line.pop();
+                    // TODO set delim for later
+                } else {
+                    // TODO set delim for later
+                }
+            }
+        } else {
+            // Possibly fragmented data, add it to buffer
+            self.tc_buf.extend_from_slice(input);
+            return -1
+        }
+        0
+    }
+}
+
+
+pub unsafe extern "C" fn rs_smtp_call_getline(input: *const u8, input_len: u32, dir: u8) -> i8 {
+    if input.is_null() && input_len > 0 {
+        return -1;
+    }
+    let buf = build_slice!(input, input_len as usize);
+    let mut smtp_buffer = RustSMTPBuffer::new();
+    match dir.into() {
+        Direction::ToServer => {
+            smtp_buffer.get_line_ts(buf)
+        }
+        Direction::ToClient => {
+            smtp_buffer.get_line_tc(buf)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
think it would be good to add a test there the input contains the end of a line, and then some more data that does not contain a line end
`EHLO [192.168.0.158]<CR><LF>more data`

Also another test to test multiple lines in one block
```
COMMAND1 data<CR><LF>
COMMAND2 data<CR><LF>
COMMAND3 data<CR><LF>
```
etc

Reply:
OK. Coming up!